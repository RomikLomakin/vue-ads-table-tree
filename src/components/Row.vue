<template>
    <tr
        :class="rowClasses"
    >
        <vue-ads-cell
            v-for="(column, key) in columns.items"
            slot="slot(column)"
            :key="key"
            :row-index="$vnode.key"
            :row="row"
            :column="column"
            :classes="classes"
            @toggleChildren="$emit('toggleChildren');"
        />
    </tr>
</template>

<script>
import Row from '../models/Row';
import VueAdsCell from './Cell';

import ClassProcessor from '../services/ClassProcessor';

export default {
    name: 'VueAdsRow',

    components: {
        VueAdsCell,
    },

    props: {
        row: {
            type: Row,
            required: true,
        },

        columns: {
            type: Object,
            required: true,
        },

        slots: {
            type: Object,
            required: false,
            default: () => { return {}; },
        },

        classes: {
            type: ClassProcessor,
            required: true,
        },
    },

    computed: {
        rowClasses () {
            return Object.assign(
                this.classes.process(this.$vnode.key + 1, null, this.row),
                this.row.classes ? ClassProcessor.processValue(this.row.classes.row, this.row) : {}
            );
        },
    },

    methods: {
        columnSlot (column) {
            return this.slots[column.property + '_' + this.row[column.property]] ||
                this.slots[column.property] ||
                null;
        },
    },
};
</script>
