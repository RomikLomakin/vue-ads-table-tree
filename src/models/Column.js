export default class Column {
    constructor (properties) {
        Object.assign(this, properties);
    }

    set property (property) {
        this._property = property;
    }

    get property () {
        if (!this.hasOwnProperty('_property')) {
            this._property = '';
        }

        return this._property;
    }

    set order (order) {
        this._order = order;
    }

    get order () {
        if (!this.hasOwnProperty('_order')) {
            return 0;
        }

        return this._order;
    }

    set collapseIcon (collapseIcon) {
        this._collapseIcon = collapseIcon;
    }

    get collapseIcon () {
        if (!this.hasOwnProperty('_collapseIcon')) {
            this._collapseIcon = false;
        }

        return this._collapseIcon;
    }

    set sortable (sortable) {
        this._sortable = sortable;
    }

    get sortable () {
        if (!this.hasOwnProperty('_sortable')) {
            this._sortable = false;
        }

        return this._sortable;
    }

    set direction (direction) {
        if (![true, false, null].includes(direction)) {
            throw new Error(
                'Sort direction has to be true (asc), false (desc) or null (no sorting)',
            );
        }

        this._direction = direction;
    }

    get direction () {
        if (!this.hasOwnProperty('_direction')) {
            this._direction = null;
        }

        return this._direction;
    }

    set filterable (filterable) {
        this._filterable = filterable;
    }

    get filterable () {
        if (!this.hasOwnProperty('_filterable')) {
            this._filterable = false;
        }

        return this._filterable;
    }

    sort (order) {
        this.direction = !this.direction;
        this.order = order;
    }
}
