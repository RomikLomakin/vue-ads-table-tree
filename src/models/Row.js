import RowCollection from '../collections/RowCollection';

export default class Row {
    constructor (properties) {
        this.children = new RowCollection();
        this.processedChildren = null;
        this.showChildren = false;
        this.hasChildren = false;
        this.childrenLoading = false;
        Object.assign(this, properties);
    }

    set parent (parent) {
        if (!(parent instanceof Row)) {
            throw new Error(
                'Parent of a row has to be a row'
            );
        }

        this._parent = parent;
    }

    get parent () {
        return this._parent;
    }

    set children (children) {
        this._children = children instanceof RowCollection ? children : new RowCollection(children);
        this._children.items
            .forEach(childRow => {
                childRow.parent = this;
            });

        this.hasChildren = !this._children.isEmpty();
    }

    get children () {
        return this._children;
    }

    set childrenLoading (childrenLoading) {
        this._childrenLoading = childrenLoading;
    }

    get childrenLoading () {
        return this._childrenLoading;
    }

    set processedChildren (processedChildren) {
        this._processedChildren = processedChildren;

        // wss voor async calls.
        // this._processedChildren.items
        //     .forEach(filteredChild => {
        //         filteredChild.parent = this;
        //     });
    }

    get processedChildren () {
        return this._processedChildren === null ? this.children : this._processedChildren;
    }

    set hasChildren (hasChildren) {
        this._hasChildren = hasChildren;
    }

    get hasChildren () {
        return this._hasChildren;
    }

    set showChildren (showChildren) {
        this._showChildren = showChildren;
    }

    get showChildren () {
        return this._showChildren;
    }

    get properties () {
        return Object.getOwnPropertyNames(this)
            .filter(property => {
                return property[0] !== '_';
            });
    }

    toggleChildren () {
        this.showChildren = !this.showChildren;
    }

    childrenLoaded () {
        return this.hasChildren && !this.children.isEmpty();
    }

    loadChildren () {
        return this.showChildren && this.hasChildren && this.children.isEmpty();
    }

    countParents () {
        if (this.parent) {
            return this.parent.countParents() + 1;
        }

        return 0;
    }
}
