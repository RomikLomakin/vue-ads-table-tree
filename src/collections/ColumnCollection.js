import Column from '../models/Column';
import AbstractCollection from './AbstractCollection';

export default class ColumnCollection extends AbstractCollection {
    constructor (items = []) {
        super(items);

        if (!this.hasCollapseIcon() && !this.empty()) {
            this.first.collapseIcon = true;
        }
    }

    get properties () {
        return this.items.map(column => column.property);
    }

    get filterColumnNames () {
        return this.items
            .filter(column => column.filterable)
            .map(column => column.property);
    }

    push (items, startIndex) {
        if (!Array.isArray(items)) {
            return super.push(items, startIndex);
        }

        return super.push(
            items.map(item => {
                if (item instanceof Column) {
                    return item;
                }

                return new Column(item);
            }),
            startIndex
        );
    }

    sortableColumns () {
        return this.items
            .filter(column => column.sortable);
    }

    sortColumns () {
        return this.sortableColumns()
            .filter(column => column.direction !== null)
            .sort((columnA, columnB) => columnA.order - columnB.order);
    }

    sort (column) {
        column.sort(this.maxOrder() + 1);
    }

    maxOrder () {
        let sortColumns = this.sortColumns();

        if (sortColumns.length === 0) {
            return 0;
        }

        return sortColumns[sortColumns.length - 1].order;
    }

    hasFilterColumns () {
        return this.filterColumnProperties.length > 0;
    }

    hasCollapseIcon () {
        return this.items.find(column => column.collapseIcon);
    }
}
