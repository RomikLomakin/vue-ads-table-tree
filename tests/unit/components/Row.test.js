import { shallowMount } from '@vue/test-utils';

import RowComponent from '../../../src/components/Row';
import Row from '../../../src/models/Row';
import Column from '../../../src/models/Column';
import ColumnCollection from '../../../src/collections/ColumnCollection';
import ClassesProcessor from '../../../src/services/ClassProcessor';

describe('Row', () => {
    let component;
    let row;
    let columns;
    let columnA;
    let columnB;


    beforeEach(() => {
        row = new Row({
            firstName: 'Arne',
            lastName: 'De Smedt',
        });

        columnA = new Column({
            property: 'firstName',
        });
        columnB = new Column({
            property: 'lastName',
        });

        columns = new ColumnCollection([
            columnA,
            columnB,
        ]);

        component = shallowMount(RowComponent, {
            propsData: {
                row,
                columns,
                rowIndex: 0,
                classes: new ClassesProcessor({}, 0),
            },
        });
    });

    it('renders a row', () => {
        expect(component.html()).toMatchSnapshot();
    });

    it('selects the correct column slot', () => {
        component.setProps({
            slots: {
                firstName: 'test',
                lastName: 'test2',
            },
        });

        expect(component.vm.columnSlot(columnA)).toBe('test');
    });

    it('selects the correct cell slot', () => {
        component.setProps({
            slots: {
                firstName_Arne: 'testcell',
                firstName: 'test',
            },
        });

        expect(component.vm.columnSlot(columnA)).toBe('testcell');
    });
});
