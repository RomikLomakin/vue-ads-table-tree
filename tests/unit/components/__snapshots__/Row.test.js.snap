// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row renders a row 1`] = `
<tr class="">
  <vueadscell-stub row="[object Object]" column="[object Object]" classes="[object Object]"></vueadscell-stub>
  <vueadscell-stub row="[object Object]" column="[object Object]" classes="[object Object]"></vueadscell-stub>
</tr>
`;
